### PR TITLE
Kmalloc and Kfree

### PR DESCRIPTION
**Pull request by rboccabe**
kmalloc and kfree are now available for allocating memory in kernel mode and releasing previously allocated memory in kernel mode. This can be tested by using  the following lines before the while(1) loop in main:

```
void *ptr1 = kmalloc(26);
void *ptr2 = kmalloc(13);
void *ptr3 = kmalloc(35);
kfree(ptr2);
ptr2 = kmalloc(4);
void *ptr4 = kmalloc(39);
void *ptr5 = kmalloc(2);
console_printf("%d for 26 -> 32\n%d for 4 -> 8\n%d for 35 -> 40\n%d for 39 -> 48\n%d for 2 -> 8\n", (int)ptr1, (int)ptr2, (int)ptr3, (int)ptr4, (int)ptr5);
```

### CODE REVIEW

File: src/memory.c [hunk 1]
```diff
@@ -4,12 +4,15 @@ This software is distributed under the GNU General Public License.
...
 
+#include "process.h"
 #include "console.h"
 #include "kerneltypes.h"
 #include "memory.h"
 #include "string.h"
 #include "memorylayout.h"
 #include "kernelcore.h"
+#include "pagetable.h"
+#include "memory.h"
```
Comment (shuyangli) [inline src/memory.c]: This is redundant

File: src/memory.c [hunk 2]
```diff
@@ -109,3 +116,196 @@ void memory_free_page(void *pageaddr) {
...
+    int i;
+    for (i = 0; i < KMALLOC_NUM_SLOTS; i++) {
+        if (page_info->free[i] == 1) {
+            consecutive_safe++;
+            console_printf("cs=%d:i=%d,",consecutive_safe,i);
+            if (consecutive_safe == num_slots) {
+                return gaps_first_slot;
+            }
+        }
+        else {
```
Comment (shuyangli) [inline src/memory.c]: style: this should be `} else {`

File: src/memory.c [hunk 3]
```diff
@@ -109,3 +116,196 @@ void memory_free_page(void *pageaddr) {
...
+    int biggest_gap_in_slots = 0;
+    int current_gap_in_slots = 0;
+    for (i = 0; i < KMALLOC_NUM_SLOTS; i++) {
+        if (page_info->free[i]) {
+            current_gap_in_slots++;
+            if (current_gap_in_slots > biggest_gap_in_slots) {
+                biggest_gap_in_slots = current_gap_in_slots;
+            }
+        }
+        else {
```
Comment (shuyangli) [inline src/memory.c]: style: this should be `} else {`

File: src/memory.c [hunk 4]
```diff
@@ -109,3 +116,196 @@ void memory_free_page(void *pageaddr) {
...
+            page_info = page_info->next;
+        }
+    }
+
+    if (is_freed == 0) {
+        console_printf("kfree(%x) failed as %x was not kmalloc'd\n");
+        //TODO: raise error?
+    }
+    else {
```
Comment (shuyangli) [inline src/memory.c]: style: this should be `} else {`

File: src/memory.c [hunk 5]
```diff
@@ -109,3 +116,196 @@ void memory_free_page(void *pageaddr) {
...
+        uint32_t page_start = (uint32_t)page_info;
+
+        //not <= and >= because pointers should never be on first or last byte of page
+        if ( page_start < to_free && page_start + PAGE_SIZE > to_free) {
+            //liberate pointer from page
+            kfree_mark_free(page_info, to_free);
+
+            is_freed = 1;
+        }
+        else {
```
Comment (shuyangli) [inline src/memory.c]: style: this should be `} else {`

File: src/memory.c [hunk 6]
```diff
@@ -109,3 +116,196 @@ void memory_free_page(void *pageaddr) {
...
+    }
+
+    uint32_t to_free_as_int = (uint32_t)to_free;
+    struct kmalloc_page_info *page_info = kmalloc_head;
+    int is_freed = 0;
+    while (page_info && is_freed == 0) {
+        uint32_t page_start = (uint32_t)page_info;
+
+        //not <= and >= because pointers should never be on first or last byte of page
+        if ( page_start < to_free && page_start + PAGE_SIZE > to_free) {
```
Comment (shuyangli) [inline src/memory.c]: style: please get rid of the space

File: src/memory.c [hunk 7]
```diff
@@ -109,3 +116,196 @@ void memory_free_page(void *pageaddr) {
     freemap[cellnumber] |= cellmask;
     pages_free++;
 }
+
+struct kmalloc_page_info * kmalloc_create_page_info(unsigned paddr) {
```
Comment (shuyangli) [inline src/memory.c]: style: please get rid of the space between `*` and `kmalloc_create_page_info`

File: src/memory.c (hunk 2 above)
Comment (rboccabe) [inline src/memory.c]: The style guide says closing braces should be on their own line?

**Comment by kkoser** [comment]
Can we change the commit message so that it describes how the code actually works? The goal of the commit message is to describe the purpose, implementation, and testing of the feature.

File: src/memory.c [hunk 8]
```diff
@@ -109,3 +116,196 @@ void memory_free_page(void *pageaddr) {
...
+    struct kmalloc_page_info *n = (struct kmalloc_page_info *)paddr;
+    n->max_free_gap = KMALLOC_NUM_SLOTS;
+    memset(n->free, 1, KMALLOC_SLOT_SIZE * KMALLOC_NUM_SLOTS);
+    return (struct kmalloc_page_info *)n;
```
Comment (kkoser) [inline src/memory.c]: cast is unnecessary

File: src/memory.c [hunk 9]
```diff
@@ -109,3 +116,196 @@ void memory_free_page(void *pageaddr) {
...
+}
+
+void kmalloc_get_page() {
+    //get current process's pagetable
+    struct pagetable *pt = current->pagetable;
+    unsigned vaddr = kmalloc_next_vaddr;
```
Comment (kkoser) [inline src/memory.c]: I don't think this var is necessary anymore. Half the time in this function we used it, half the time we just used kmalloc_next_vaddr. I vote we remove it.

File: src/memory.c [hunk 10]
```diff
@@ -109,3 +116,196 @@ void memory_free_page(void *pageaddr) {
...
+void *kmalloc(unsigned int size)
+{
+    //console_printf("In kmalloc with requested size: %d\n", size);
+    uint16_t slots_needed = (size + sizeof(uint16_t)) / KMALLOC_SLOT_SIZE;
+    //addresses integer division truncation
+    if (size % 8 != 0) {
+        slots_needed++;
+    }
+    
+    console_printf("For a request of size %d we need a gap of at least %d slots\n", size, slots_needed);
```
Comment (kkoser) [inline src/memory.c]: remove this (and other logs) before merging plz.

File: src/memory.c [hunk 11]
```diff
@@ -109,3 +116,196 @@ void memory_free_page(void *pageaddr) {
...
+
+    //start at head, iterate through looking for a page with a large enough gap
+    struct kmalloc_page_info *page_info = kmalloc_head;
+    while (page_info) {
+        console_printf("Checking out page %x\n", (int)page_info);
+        if (page_info->max_free_gap >= slots_needed) {
+            console_printf("suitable page found with max gap of %d\n", page_info->max_free_gap);
```
Comment (kkoser) [inline src/memory.c]: remove

File: src/memory.c [hunk 12]
```diff
@@ -109,3 +116,196 @@ void memory_free_page(void *pageaddr) {
...
+    if (size % 8 != 0) {
+        slots_needed++;
+    }
+    
+    console_printf("For a request of size %d we need a gap of at least %d slots\n", size, slots_needed);
+
+    //start at head, iterate through looking for a page with a large enough gap
+    struct kmalloc_page_info *page_info = kmalloc_head;
+    while (page_info) {
+        console_printf("Checking out page %x\n", (int)page_info);
```
Comment (kkoser) [inline src/memory.c]: remove

File: src/memory.c [hunk 13]
```diff
@@ -109,3 +116,196 @@ void memory_free_page(void *pageaddr) {
...
+        if (page_info->max_free_gap >= slots_needed) {
+            console_printf("suitable page found with max gap of %d\n", page_info->max_free_gap);
+            break; 
+        }
+        console_printf("unsuitable page found with max gap of %d\n", page_info->max_free_gap);
```
Comment (kkoser) [inline src/memory.c]: remove

File: src/memory.c [hunk 14]
```diff
@@ -109,3 +116,196 @@ void memory_free_page(void *pageaddr) {
...
+        page_info = page_info->next;
+    }
+
+    //If no such page exists, grab another!
+    if (!page_info) {
+        console_printf("no suitable pages found, getting a new page\n");
```
Comment (kkoser) [inline src/memory.c]: remove

File: src/memory.c [hunk 15]
```diff
@@ -109,3 +116,196 @@ void memory_free_page(void *pageaddr) {
...
+        kmalloc_get_page();
+        page_info = kmalloc_head;
+    }
+
+    //find some large enough gap
+    int slots_start_index = kmalloc_locate_sufficient_gap(page_info, slots_needed);
+    console_printf("slots_start_index = %d\n", slots_start_index);
```
Comment (kkoser) [inline src/memory.c]: remove

File: src/memory.c [hunk 16]
```diff
@@ -109,3 +116,196 @@ void memory_free_page(void *pageaddr) {
...
+    //mark these as malloc'd in kmalloc_page
+    int i;
+    for (i = 0; i < slots_needed; i++) {
+        page_info->free[slots_start_index + i] = 0;
+
+    }
+    
+    //find largest remaining gap
+    page_info->max_free_gap = kmalloc_get_largest_gap_size(page_info);
+    console_printf("\nLargest remaining gap is %d slots\n", page_info->max_free_gap);
```
Comment (kkoser) [inline src/memory.c]: remove

File: src/memory.c [hunk 17]
```diff
@@ -109,3 +116,196 @@ void memory_free_page(void *pageaddr) {
...
+
+    //find some large enough gap
+    int slots_start_index = kmalloc_locate_sufficient_gap(page_info, slots_needed);
+    console_printf("slots_start_index = %d\n", slots_start_index);
+     
+    //fill in the front of the first slot with number of slots info
+    //phys addr is addr of page info plus size of page info plus slot offset * slot size
+    //needs to fit a memory address of 4 bytes
+    uint32_t first_slot_phys_addr = (uint32_t)page_info + sizeof(struct kmalloc_page_info) + (KMALLOC_SLOT_SIZE * slots_start_index);
+    console_printf("Found slot with first_slot_phys_addr = %d\n", first_slot_phys_addr);
```
Comment (kkoser) [inline src/memory.c]: remove

File: src/memory.c [hunk 18]
```diff
@@ -109,3 +116,196 @@ void memory_free_page(void *pageaddr) {
...
+    console_printf("\nLargest remaining gap is %d slots\n", page_info->max_free_gap);
+
+    //return a pointer to the useful memory area, which is sizeof(uint16_t) into the first slot.
+    return (void *)(first_slot_phys_addr + sizeof(uint16_t));
+}
+
+void kfree_mark_free(struct kmalloc_page_info *page_info, void *mem_loc) {
+    //the number of consecutive slots consumed is stored just ahead of the pointer given
+    uint16_t slots_consumed = *((uint16_t *)(mem_loc - sizeof(uint16_t)));
+    console_printf("Freeing %d blocks beginning at mem_loc %x\n", slots_consumed, mem_loc - sizeof(uint16_t));
```
Comment (kkoser) [inline src/memory.c]: remove

File: src/memory.c (hunk 6 above)
Comment (kkoser) [inline src/memory.c]: I think we need casts here, or warnings will be thrown about using pointers for arithmetic.

File: src/memory.c [hunk 19]
```diff
@@ -109,3 +116,196 @@ void memory_free_page(void *pageaddr) {
...
+    
+    //find largest remaining gap
+    page_info->max_free_gap = kmalloc_get_largest_gap_size(page_info);
+    console_printf("\nLargest remaining gap is %d slots\n", page_info->max_free_gap);
+
+    //return a pointer to the useful memory area, which is sizeof(uint16_t) into the first slot.
+    return (void *)(first_slot_phys_addr + sizeof(uint16_t));
+}
+
+void kfree_mark_free(struct kmalloc_page_info *page_info, void *mem_loc) {
```
Comment (kkoser) [inline src/memory.c:262]: Comments for function describing purpose?

File: src/memory.c [hunk 20]
```diff
@@ -109,3 +116,196 @@ void memory_free_page(void *pageaddr) {
...
+    }
+
+    //Return available number of slots in the largest gap.
+    //Everything regarding the space in the front to track the number of slots
+    //is taken care of in kmalloc proper.
+    return biggest_gap_in_slots;
+}
+
+void *kmalloc(unsigned int size)
+{
```
Comment (kkoser) [inline src/memory.c]: style: bracket placement

File: src/memory.c [hunk 21]
```diff
@@ -109,3 +116,196 @@ void memory_free_page(void *pageaddr) {
...
+    // build the page struct
+    struct kmalloc_page_info *pg_info = kmalloc_create_page_info(phys_addr);
+    
+ 
+    //rearrange the linked list of pages with the new page at the beginning.
+    pg_info->next = kmalloc_head;
+    kmalloc_head = pg_info;
+}
+
+int kmalloc_locate_sufficient_gap(struct kmalloc_page_info *page_info, int num_slots) {
```
Comment (kkoser) [inline src/memory.c:165]: Comments for function purpose?

File: src/memory.c [hunk 22]
```diff
@@ -109,3 +116,196 @@ void memory_free_page(void *pageaddr) {
...
 }
+
+struct kmalloc_page_info * kmalloc_create_page_info(unsigned paddr) {
+    struct kmalloc_page_info *n = (struct kmalloc_page_info *)paddr;
+    n->max_free_gap = KMALLOC_NUM_SLOTS;
+    memset(n->free, 1, KMALLOC_SLOT_SIZE * KMALLOC_NUM_SLOTS);
+    return (struct kmalloc_page_info *)n;
+}
+
+void kmalloc_get_page() {
```
Comment (kkoser) [inline src/memory.c:130]: Comments for function?

File: src/main.c [hunk 23]
```diff
@@ -45,9 +45,8 @@ int kernel_main() {
 
     ata_init();
 
-
-    console_printf("\nBASEKERNEL READY:\n");
-
+	console_printf("\nBASEKERNEL READY:\n");
```
Comment (alexmwu) [inline src/main.c]: tab

File: src/memory.c (hunk 1 above)
Comment (rboccabe) [inline src/memory.c]: addressed in an upcoming commit.

File: src/memory.c (hunk 3 above)
Comment (rboccabe) [inline src/memory.c]: no fix.

File: src/memory.c (hunk 4 above)
Comment (rboccabe) [inline src/memory.c]: no fix

File: src/memory.c (hunk 5 above)
Comment (rboccabe) [inline src/memory.c]: no fix

File: src/memory.c (hunk 6 above)
Comment (rboccabe) [inline src/memory.c]: both addressed in upcoming code review commit.

File: src/memory.c (hunk 7 above)
Comment (rboccabe) [inline src/memory.c]: addressed in upcoming code review commit.

File: src/memory.c (hunk 8 above)
Comment (rboccabe) [inline src/memory.c]: addressed in upcoming code review commit.

File: src/memory.c (hunk 9 above)
Comment (rboccabe) [inline src/memory.c]: Good call. It was useful for seeing how the += PAGE_SIZE was affecting things during a debug print statement, but is unnecessary now. taken care of in upcoming commit.

File: src/memory.c (hunk 20 above)
Comment (rboccabe) [inline src/memory.c]: addressed in upcoming code review commit.

File: src/main.c (hunk 23 above)
Comment (rboccabe) [inline src/main.c]: addressed in upcoming code review commit.

File: src/memory.c (hunk 12 above)
Comment (rboccabe) [inline src/memory.c]: all of these removes were on an outdated diff... I'm not going to address the rest of them.

File: src/memory.c (hunk 10 above)
Comment (rboccabe) [inline src/memory.c]: commented on outdated diff

File: src/memory.c (hunk 11 above)
Comment (rboccabe) [inline src/memory.c]: commented on outdated diff

File: src/memory.c (hunk 21 above)
Comment (rboccabe) [inline src/memory.c:165]: @kkoser Do these follow the Doxygen standard we set out for the .h files? Or is it just a multiline comment that describes how it helps.

File: src/memory.h [hunk 24]
```diff
@@ -9,8 +9,39 @@ See the file LICENSE for details.
...
+    uint8_t free[KMALLOC_NUM_SLOTS]; // bit vector of free slots
+};
+
 void memory_init();
 void *memory_alloc_page(bool zeroit);
 void memory_free_page(void *addr);
 
+/**
+* @brief Kernel allocation of the parameter requested size of memory
+* @details Looks goes through a linked list of pages for a large enough gap of contiguous free slots of fixed size . If no page for kmalloc has a large enough gap, a new gap is asked from memory.
```
Comment (zachwaterson) [inline src/memory.h]: "Looks goes"? We should fix the wording here. Also the space before and after the period: "size . If"

Comment (rboccabe) [inline src/memory.h]: addressed in upcoming code review commit.

File: src/memory.c (hunk 2 above)
Comment (kkoser) [inline src/memory.c]: Not for else's: "When “if” statements have “else if” or “else” branches, put them on the same line as the previous closing brace"

Comment (rboccabe) [inline src/memory.c]: addressed in upcoming code review commit.

File: src/memory.c (hunk 3 above)
Comment (rboccabe) [inline src/memory.c]: addressed in upcoming code review commit.

File: src/memory.c (hunk 4 above)
Comment (rboccabe) [inline src/memory.c]: addressed in upcoming code review commit.

File: src/memory.c (hunk 5 above)
Comment (rboccabe) [inline src/memory.c]: addressed in upcoming code review commit.

File: src/memory.c (hunk 19 above)
Comment (rboccabe) [inline src/memory.c:262]: addressed in upcoming code review commit.

File: src/memory.c (hunk 21 above)
Comment (rboccabe) [inline src/memory.c:165]: addressed in upcoming code review commit.

File: src/memory.c (hunk 22 above)
Comment (rboccabe) [inline src/memory.c:130]: addressed in upcoming code review commit.

File: src/memory.c [hunk 25]
```diff
@@ -109,3 +115,196 @@ void memory_free_page(void *pageaddr) {
...
+
+/**
+* @brief Returns the largest gap in terms of the number of slots in that gap.
+* @details Looks through the arugments free array, looking for the longest consecutive string of free slots
+*
+* @param page_info Pointer to the kmalloc_page_info which maintains the details regarding the page we're looking for the largest gap in.
+* @return The number of slots in the largest continuous, free gap on the page.
+*/
+int kmalloc_get_largest_gap_size(struct kmalloc_page_info *page_info)
+{
```
Comment (zachwaterson) [inline src/memory.c]: Style: this should be on the previous line

File: src/memory.c [hunk 26]
```diff
@@ -109,3 +115,196 @@ void memory_free_page(void *pageaddr) {
...
+        page_info->free[first_slot_index + i] = 1;
+    }
+}
+
+void kfree(void *to_free) {
+    if (!to_free) {
+        return;
+    }
+
+    uint32_t to_free_as_int = (uint32_t)to_free;
```
Comment (zachwaterson) [inline src/memory.c]: This variable is unused. Intentional?

Comment (rboccabe) [inline src/memory.c]: addressed in upcoming code review commit.

File: src/memory.c (hunk 25 above)
Comment (rboccabe) [inline src/memory.c]: addressed in upcoming code review commit.

**Pull request merged**